File: scripts/stamp-versions.py
#!/usr/bin/env python3
"""Rewrite version fields in the pip packaging manifests.

    python3 scripts/stamp-versions.py 0.1.0

Affects:
    packaging/pip/pyproject.toml
    packaging/pip/src/sqlitedeploy/__init__.py
"""

from __future__ import annotations

import re
import sys
from pathlib import Path

SEMVER_RE = re.compile(r"^\d+\.\d+\.\d+")


def main() -> None:
    if len(sys.argv) != 2 or not SEMVER_RE.match(sys.argv[1]):
        print("usage: stamp-versions.py <semver>", file=sys.stderr)
        sys.exit(1)
    version = sys.argv[1]
    repo_root = Path(__file__).resolve().parent.parent

    pyproject = repo_root / "packaging" / "pip" / "pyproject.toml"
    text = pyproject.read_text(encoding="utf-8")
    new_text, n = re.subn(
        r'(?m)^version = "[^"]+"', f'version = "{version}"', text, count=1
    )
    if n != 1:
        raise SystemExit(f"could not find version field in {pyproject}")
    pyproject.write_text(new_text, encoding="utf-8")
    print(f"stamped {pyproject.relative_to(repo_root)} -> {version}")

    init = repo_root / "packaging" / "pip" / "src" / "sqlitedeploy" / "__init__.py"
    text = init.read_text(encoding="utf-8")
    new_text, n = re.subn(
        r'(?m)^__version__ = "[^"]+"', f'__version__ = "{version}"', text, count=1
    )
    if n != 1:
        raise SystemExit(f"could not find __version__ in {init}")
    init.write_text(new_text, encoding="utf-8")
    print(f"stamped {init.relative_to(repo_root)} -> {version}")


if __name__ == "__main__":
    main()

File: scripts/stamp-versions.js
#!/usr/bin/env node
// Rewrite version fields in all npm package manifests.
//   node scripts/stamp-versions.js 0.1.0
// Affects:
//   packaging/npm/sqlitedeploy/package.json   (main package + optionalDependencies)
//   packaging/npm/platforms/<plat>/package.json (each platform package)

'use strict';

const fs = require('node:fs');
const path = require('node:path');

const version = process.argv[2];
if (!version || !/^\d+\.\d+\.\d+/.test(version)) {
  console.error('usage: stamp-versions.js <semver>');
  process.exit(1);
}

const repoRoot = path.resolve(__dirname, '..');
const platformsDir = path.join(repoRoot, 'packaging', 'npm', 'platforms');
const mainPkgPath = path.join(repoRoot, 'packaging', 'npm', 'sqlitedeploy', 'package.json');

function rewrite(file, mutate) {
  const json = JSON.parse(fs.readFileSync(file, 'utf8'));
  mutate(json);
  fs.writeFileSync(file, JSON.stringify(json, null, 2) + '\n');
  console.log(`stamped ${path.relative(repoRoot, file)} -> ${version}`);
}

// Each platform package gets the version directly.
for (const dir of fs.readdirSync(platformsDir)) {
  const pkg = path.join(platformsDir, dir, 'package.json');
  if (!fs.existsSync(pkg)) continue;
  rewrite(pkg, (json) => {
    json.version = version;
  });
}

// Main package: bump its own version AND every entry in optionalDependencies
// to the same version, so the resolver can require them at exactly that pin.
rewrite(mainPkgPath, (json) => {
  json.version = version;
  if (json.optionalDependencies) {
    for (const dep of Object.keys(json.optionalDependencies)) {
      json.optionalDependencies[dep] = version;
    }
  }
});
